### DISCUSSION TITLE
Implement `Changed` in custom widget?

### DISCUSSION THREAD

**Discussion by TomJGooding** [Textualize/textual#2410] [Q&A] [answered]
Many of the widgets in Textual have a `Changed` message which then "can be handled... in a parent widget in the DOM" such as `on_switch_changed` - how would you implement this in a custom subclass of a `Widget`?

I've looked through the widgets codebase and the docs but I'm afraid I still don't really understand how this works. Sorry if I am missing something obvious that is explained in the docs...

For context, I thought I'd have a go at implementing a date picker widget just as a little personal project to better understand Textual. I'm struggling a bit to reduce this to a minimal repro, so hope okay to ask for help without this in case I have just missed something in the docs?

Many thanks for any advice!

**Comment by davep** [comment, marked as answer]
Does [the documentation on how to create custom messages](https://textual.textualize.io/guide/events/#custom-messages) help?

**Reply by TomJGooding** [reply]
Thanks Dave for your help as always!

I have read through this documentation a few times, but I must be missing something to replicate the same behaviour as the regular widgets in Textual.

I was looking at the `Switch` widget for a simple example hoping to understand how this works (which is when I spotted the docstring fix!)

```python3
    class Changed(Message, bubble=True):
    # --- snip ---
        def __init__(self, switch: Switch, value: bool) -> None:
            super().__init__()
            self.value: bool = value
            self.switch: Switch = switch

        @property
        def control(self) -> Switch:
            """Alias for self.switch."""
            return self.switch

**Reply by davep** [reply]
>  but I must be missing something to replicate the same behaviour as the regular widgets in Textual

Do you think you could clarify what isn't quite making sense for you, or do you perhaps have a minimal attempt at making it work that you can share so we can go over what might be wrong?

**Reply by TomJGooding** [reply]
Thanks Dave, I'll have another go at creating a minimal repro.

**Reply by TomJGooding** [reply]
Not _exactly_ sure where I was going wrong, but after some changes to match the examples in the codebase/docs I finally managed to get this working! Thanks again for your help.

**Reply by davep** [reply]
Glad you got it sorted. 🙂
